### PR TITLE
Don't `set_params` at the end of `set_state`.

### PR DESCRIPTION
**Pull request by trivialfis**
* Also fix another issue found in dask prediction.

### CODE REVIEW

File: tests/python/test_pickling.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    with open(filename, 'rb') as fd:
+        bst = pickle.load(fd)
+
+    with open(filename, 'wb') as fd:
+        pickle.dump(bst, fd)
+
+    with open(filename, 'rb') as fd:
+        bst = pickle.load(fd)
+
+    assert bst.get_dump(dump_format='json')
```
Comment (JohnZed) [inline tests/python/test_pickling.py]: should we assert that this is identical to the first model dump not just nonempty?

Comment (trivialfis) [reply tests/python/test_pickling.py]: Yup.  We have continuous training test, but adding it would be useful.

**Comment by trivialfis** [comment]
Also partially revert the previous fix.  My brain is fried ...

**Comment by RAMitchell** [comment]
For others following, this appears to be a fundamental error in python serialisation where the model can be pickled and unpickled multiple times, resulting in a different model.

**Pull request merged**